### PR TITLE
✨ Add ability to scrape all s3 object versions

### PR DESCRIPTION
**Pull request by chris-s-friedman**
- [ ] README entry added if new functionality
- [ ] fixup commits are appropriately squashed

Add functionality to be able to scrape aws s3 object versions + delete markers in addition to bucket objects.

### CODE REVIEW

File: d3b_utils/aws_bucket_contents.py [hunk 1]
```diff
@@ -81,9 +108,15 @@ def fetch_bucket_obj_info(
...
+                dict_writer.writeheader()
+                dict_writer.writerows(delete_markers)
 
+    if get_versions and delete_markers:
+        return "a dict of bucket_contents + delete markers maybe?"
     return bucket_contents
-
-
-# backwards compatibility
-fetch_aws_bucket_obj_info = fetch_bucket_obj_info
```
Comment (fiendish) [inline d3b_utils/aws_bucket_contents.py:89]: I don't see a reason to remove this alias

**Comment by fiendish** [comment]
The 'Versions' key in the response object for list_object_versions looks like it contains all the same data (and more) as the 'Contents' object of list_objects_v2 response, so, assuming the versions paginator works well, I think you shouldn't need to use both iterators and should be able to just pick one depending on the passed argument.

File: d3b_utils/aws_bucket_contents.py (hunk 1 above)
Comment (chris-s-friedman) [reply d3b_utils/aws_bucket_contents.py:89]: oh oops - did that by mistake. Thanks for catching that.

Comment (chris-s-friedman) [reply d3b_utils/aws_bucket_contents.py]: fixed

**Comment by chris-s-friedman** [comment]
I like that - i'll refactor to only give object versions if that flag is selected. 

When I was putting this together yesterday, I had not decided on how I want to handle delete markers. Since I'll be returning versions alone and not a combination of versions and bucket contents, I'm thinking that, when `get_versions = True`, the function will return a dict with both versions and delete markers in it, i.e.:

```
{
  versions: [{version-a}, {version-b}], 
  delete_markers:[{delete_marker-a}, {delete_marker-b}], 
}
```

What I still want to think through is what the behavior should be if `get_versions = True` _and_ a user specifies an `output_filename`. I think that if output_filename is specified, there should be two files returned - something like `"versions" + output_filename` and `"delete_markers" + output_filename`. 

I don't think it makes sense to return versions and delete markers in the same file as two separate columns.

**Comment by fiendish** [comment]
Interesting dilemma re file names.  Yeah what you suggest sounds ok, and I don't think it will be terrible if we just put a comment in the docstring about it.

**Comment by chris-s-friedman** [comment]
@fiendish refactored and added/modified docstrings. (updating readme now)

**Comment by chris-s-friedman** [comment]
I have been thinking about whether or not this should really be two functions:
1. a function for returning objects 
2. a function for returning versions and delete_markers

**Comment by fiendish** [comment]
I refactored the code change to remove duplicate code. Please review my edits and make sure it still does what you want.

**Comment by chris-s-friedman** [comment]
@fiendish looks great! thanks for the refactor! Updated the readme to reflect changing the arg to `all_versions`

**Review by fiendish — APPROVED** [review]
🚀

**Pull request merged**